# Bag of Words (BoW), TF-IDF, and Document Similarity

## Overview

This notebook introduces two fundamental text representation techniques used in Natural Language Processing (NLP):

- Bag of Words (BoW)
- Term Frequency-Inverse Document Frequency (TF-IDF)

These techniques convert text into numerical representations that machine learning algorithms can process and analyze.

The notebook also demonstrates how to measure similarity between documents using TF-IDF vectors and Cosine Similarity.

---

## What is Bag of Words (BoW)?

Bag of Words (BoW) is a text representation technique that converts text into numerical vectors based on word frequency.

Key characteristics:

- Creates a vocabulary of unique words
- Counts how many times each word appears
- Ignores grammar and word order
- Represents documents as numerical vectors

### Example

Document 1:
```
I love NLP
```

Document 2:
```
I love Python
```

Vocabulary:
```
[I, love, NLP, Python]
```

Vector Representation:

| Document | I | love | NLP | Python |
|-----------|---|------|-----|--------|
| Doc 1 | 1 | 1 | 1 | 0 |
| Doc 2 | 1 | 1 | 0 | 1 |

---

## Advantages of BoW

- Simple and easy to understand
- Fast to implement
- Works well for basic text classification tasks

---

## Limitations of BoW

- Ignores word order
- Does not capture meaning
- Treats all words equally
- Common words may dominate the representation

---

In [1]:
# Import required libraries
import pandas as pd

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

In [2]:
# Sample text documents
documents = [
    "I love natural language processing",
    "Natural language processing is amazing",
    "I enjoy learning machine learning",
    "Machine learning is part of artificial intelligence"
]

# Print documents
for i, doc in enumerate(documents, 1):
    print(f"Document {i}: {doc}")

Document 1: I love natural language processing
Document 2: Natural language processing is amazing
Document 3: I enjoy learning machine learning
Document 4: Machine learning is part of artificial intelligence


Bag of Words (CountVectorizer)

In [4]:
# Create BoW model
bow_vectorizer = CountVectorizer()

# Fit and transform documents
bow_matrix = bow_vectorizer.fit_transform(documents)

# Convert to dataframe for better visualization
bow_df = pd.DataFrame(
    bow_matrix.toarray(),
    columns=bow_vectorizer.get_feature_names_out()
)

print("Bag of Words Matrix:")
bow_df

Bag of Words Matrix:


,amazing,artificial,enjoy,intelligence,is,language,learning,love,machine,natural,of,part,processing
0,0,0,0,0,0,1,0,1,0,1,0,0,1
1,1,0,0,0,1,1,0,0,0,1,0,0,1
2,0,0,1,0,0,0,2,0,1,0,0,0,0
3,0,1,0,1,1,0,1,0,1,0,1,1,0


## What is TF-IDF?

TF-IDF stands for:

- TF = Term Frequency
- IDF = Inverse Document Frequency


### 1. Term Frequency (TF)
Measures how often a word appears in a document.

Example:
If a word appears many times → TF is high

---

### 2. Inverse Document Frequency (IDF)
Measures how important a word is across all documents.

- Common words (e.g., "is", "the") → low importance  
- Rare words (e.g., "NLP") → high importance  


TF-IDF improves upon BoW by assigning importance scores to words.

Words that appear frequently in a document but rarely across other documents receive higher weights.

---

## Why TF-IDF?

Not all words are equally important.

For example:

```
the, is, and
```

appear in many documents and provide little useful information.

TF-IDF reduces the importance of such common words and highlights more meaningful terms.

---

## Advantages of TF-IDF

- Considers word importance
- Reduces influence of common words
- Produces better document representations
- Widely used in information retrieval systems

---

TF-IDF Vectorization

In [5]:
# Create TF-IDF model
tfidf_vectorizer = TfidfVectorizer()

# Transform documents
tfidf_matrix = tfidf_vectorizer.fit_transform(documents)

# Convert to dataframe
tfidf_df = pd.DataFrame(
    tfidf_matrix.toarray(),
    columns=tfidf_vectorizer.get_feature_names_out()
)

print("TF-IDF Matrix:")
tfidf_df

TF-IDF Matrix:


,amazing,artificial,enjoy,intelligence,is,language,learning,love,machine,natural,of,part,processing
0,0.000000,0.000000,0.000000,0.000000,0.000000,0.465809,0.000000,0.590819,0.000000,0.465809,0.000000,0.000000,0.465809
1,0.535566,0.000000,0.000000,0.000000,0.422247,0.422247,0.000000,0.000000,0.000000,0.422247,0.000000,0.000000,0.422247
2,0.000000,0.000000,0.493386,0.000000,0.000000,0.000000,0.777982,0.000000,0.388991,0.000000,0.000000,0.000000,0.000000
3,0.000000,0.412928,0.000000,0.412928,0.325557,0.000000,0.325557,0.000000,0.325557,0.000000,0.412928,0.412928,0.000000


Show Vocabulary

In [11]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()

X = vectorizer.fit_transform(documents)

# Show vocabulary
print("Vocabulary (BoW features):")
print(vectorizer.get_feature_names_out())

# Show vocabulary size
vocab_size = len(vectorizer.get_feature_names_out())

print("Vocabulary Size:", vocab_size)

Vocabulary (BoW features):
['amazing' 'artificial' 'enjoy' 'intelligence' 'is' 'language' 'learning'
 'love' 'machine' 'natural' 'of' 'part' 'processing']
Vocabulary Size: 13


In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()

X_tfidf = tfidf.fit_transform(documents)

feature_names = tfidf.get_feature_names_out()

print("TF-IDF Features:")
print(feature_names)

TF-IDF Features:
['amazing' 'artificial' 'enjoy' 'intelligence' 'is' 'language' 'learning'
 'love' 'machine' 'natural' 'of' 'part' 'processing']


Stopwords in BoW/TF-IDF

In [12]:
from sklearn.feature_extraction.text import CountVectorizer

# Define stopwords removal inside vectorizer
vectorizer = CountVectorizer(stop_words='english')

X = vectorizer.fit_transform(documents)

print("After Stopwords Removal:")
print(vectorizer.get_feature_names_out())

After Stopwords Removal:
['amazing' 'artificial' 'enjoy' 'intelligence' 'language' 'learning'
 'love' 'machine' 'natural' 'processing']


## Sparsity Problem

BoW and TF-IDF create large matrices where:
- Most values are 0

This is called a sparse matrix.

Why it happens:
- Large vocabulary size
- Many words do not appear in every document

Impact:
- High memory usage
- Slower processing (in large datasets)

---

Sparsity Example

In [13]:
import pandas as pd

# Convert BoW matrix to dataframe
df = pd.DataFrame(X.toarray(), columns=vectorizer.get_feature_names_out())

print("BoW Matrix (Sparse Representation):")
print(df)

# Count zeros
zero_count = (df == 0).sum().sum()

print("\nNumber of zero values (sparsity):", zero_count)

BoW Matrix (Sparse Representation):
   amazing  artificial  enjoy  intelligence  language  learning  love  \
0        0           0      0             0         1         0     1   
1        1           0      0             0         1         0     0   
2        0           0      1             0         0         2     0   
3        0           1      0             1         0         1     0   

   machine  natural  processing  
0        0        1           1  
1        0        1           1  
2        1        0           0  
3        1        0           0  

Number of zero values (sparsity): 25


TF-IDF Importance Score

In [14]:
import numpy as np

# Get feature names
features = tfidf.get_feature_names_out()

# Convert matrix to array
matrix = X_tfidf.toarray()

print("Word Importance in each document:\n")

for i, doc in enumerate(matrix):
    print(f"\nDocument {i+1}:")
    
    # Get top 3 important words
    top_indices = doc.argsort()[-3:][::-1]
    
    for idx in top_indices:
        print(features[idx], ":", round(doc[idx], 3))

Word Importance in each document:


Document 1:
love : 0.591
natural : 0.466
processing : 0.466

Document 2:
amazing : 0.536
natural : 0.422
language : 0.422

Document 3:
learning : 0.778
enjoy : 0.493
machine : 0.389

Document 4:
part : 0.413
of : 0.413
artificial : 0.413


In [15]:
bow = CountVectorizer().fit_transform(documents)
tfidf = TfidfVectorizer().fit_transform(documents)

print("BoW Matrix:\n", bow.toarray())
print("\nTF-IDF Matrix:\n", tfidf.toarray())

print("\nObservation:")
print("TF-IDF reduces importance of common words and highlights important ones.")

BoW Matrix:
 [[0 0 0 0 0 1 0 1 0 1 0 0 1]
 [1 0 0 0 1 1 0 0 0 1 0 0 1]
 [0 0 1 0 0 0 2 0 1 0 0 0 0]
 [0 1 0 1 1 0 1 0 1 0 1 1 0]]

TF-IDF Matrix:
 [[0.         0.         0.         0.         0.         0.46580855
  0.         0.59081908 0.         0.46580855 0.         0.
  0.46580855]
 [0.53556627 0.         0.         0.         0.4222466  0.4222466
  0.         0.         0.         0.4222466  0.         0.
  0.4222466 ]
 [0.         0.         0.49338588 0.         0.         0.
  0.7779822  0.         0.3889911  0.         0.         0.
  0.        ]
 [0.         0.41292788 0.         0.41292788 0.32555709 0.
  0.32555709 0.         0.32555709 0.         0.41292788 0.41292788
  0.        ]]

Observation:
TF-IDF reduces importance of common words and highlights important ones.


#### Key Evidence from our Matrices:
* Unique Word Boost: In Document 0, two words both have a BoW count of `1`. However, the word unique only to Document 0 gets a higher TF-IDF score (`0.59`) than the word shared with Document 1 (`0.46`).
* Length Normalization: TF-IDF automatically scales scores based on document length, preventing longer documents from artificially dominating the feature space.
* Peak Importance: The highest score in the dataset (`0.77` in Document 2) occurs where a word has both a high repetition within that document *and* high scarcity across the collection.

## Bag of Words vs TF-IDF

| Feature | BoW | TF-IDF |
|----------|----------|----------|
| Uses Word Counts | Yes | No |
| Considers Importance | No | Yes |
| Handles Common Words Better | No | Yes |
| Information Quality | Basic | Better |
| Practical Usage | Simple Tasks | Real-world NLP |

---

## Output Analysis

After transformation:

### BoW Output:
- Matrix of word counts
- Shows frequency of each word

### TF-IDF Output:
- Weighted matrix
- Shows importance of words in documents

---

## Final Understanding

- BoW = Simple word counting approach
- TF-IDF = Advanced importance-based approach
- Both convert text into numerical data for ML models

## Real Use Case
TF-IDF is used in:
- Google search ranking
- YouTube recommendation system
- Chatbots
- Document similarity systems

# Conclusion

- Bag of Words converts text into word-frequency vectors.
- TF-IDF improves text representation by considering word importance.
- Cosine Similarity measures how closely documents are related.
- A document similarity tool can identify similar documents and is widely used in search engines, recommendation systems, and information retrieval.